mempool/src/pool/reorg.rs

```diff
@@ -0,0 +1,153 @@
+    primitives::{Id, Idable},
+};
+use logging::log;
+use utils::tap_error_log::LogError;
+use utxo::UtxosStorageRead;
+
+use crate::{get_memory_usage::GetMemoryUsage, pool::Mempool};
+
+/// An error that can happen in mempool on chain reorg
+#[derive(Debug, thiserror::Error, PartialEq, Eq)]
+pub enum ReorgError {
+    #[error(transparent)]
+    Chainstate(#[from] chainstate::ChainstateError),
+    #[error("Could not find the previous tip")]
+    OldTip,
+    #[error("Could not find the previous tip index")]
+    OldTipIndex,
+    #[error("Could not find the new tip index")]
+    NewTipIndex,
+    #[error("Block {0:?} not found while traversing history")]
+    BlockNotFound(Id<Block>),
+    #[error("Chainstate call: {0}")]
+    ChainstateCall(#[from] subsystem::subsystem::CallError),
+}
+
+/// Collect blocks between the given two points
+fn collect_blocks<C: ChainstateInterface + ?Sized>(
+    chainstate: &C,
+    mut curr_id: Id<GenBlock>,
+    stop_id: Id<GenBlock>,
+) -> Result<Vec<Block>, ReorgError> {
+    let chain_config = chainstate.get_chain_config();
+    let mut result = Vec::new();
+    while curr_id != stop_id {
+        let curr_block_id = curr_id
+            .classify(chain_config)
+            .chain_block_id()
+            .expect("Reached genesis before the stopping block");
+        let block = chainstate
+            .get_block(curr_block_id)?
+            .ok_or_else(|| ReorgError::BlockNotFound(curr_block_id))?;
+        curr_id = block.prev_block_id();
+        result.push(block);
+    }
+    Ok(result)
+}
+
+/// Blocks affected by a reorg
+struct ReorgInfo {
+    disconnected: Vec<Block>,
+    connected: Vec<Block>,
+}
+
+impl ReorgInfo {
+    /// Extract blocks that have been disconnected and connected from the chainstate.
+    fn from_chainstate<C: ChainstateInterface + ?Sized>(
+        chainstate: &C,
+        old_tip_id: Id<GenBlock>,
+        new_tip_id: Id<GenBlock>,
+    ) -> Result<Self, ReorgError> {
+        let common_id = {
+            let old_index =
+                chainstate.get_gen_block_index(&old_tip_id)?.ok_or(ReorgError::OldTipIndex)?;
+            let new_index =
+                chainstate.get_gen_block_index(&new_tip_id)?.ok_or(ReorgError::NewTipIndex)?;
+            let common_index = chainstate.last_common_ancestor(&old_index, &new_index)?;
+            common_index.block_id()
+        };
+
+        Ok(Self {
+            disconnected: collect_blocks(chainstate, old_tip_id, common_id)?,
+            connected: collect_blocks(chainstate, new_tip_id, common_id)?,
+        })
+    }
+
+    /// Get transactions that have been disconnected and not reconnected
+    fn into_disconnected_transactions(self) -> impl Iterator<Item = SignedTransaction> {
+        let connected_txs: BTreeSet<_> = self
+            .connected
+            .into_iter()
+            .flat_map(|block| {
+                block.into_transactions().into_iter().map(|tx| tx.transaction().get_id())
```

Comment:
We can go nuts with rayon here and parallize this. Not necessary though in this PR.